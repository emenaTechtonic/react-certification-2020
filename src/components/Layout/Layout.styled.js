import styled from 'styled-components'

const StyledLayout = styled.div.attrs({
  className: 'layout',
})``
export default StyledLayout
