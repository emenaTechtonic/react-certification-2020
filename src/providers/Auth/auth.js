import React, { useState, useEffect, createContext, useCallback } from 'react'
import { AUTH_STORAGE_KEY, AUTH_STORAGE_USER } from '../../utils/constants'
import { storage } from '../../utils/storage'

export const AuthContext = createContext(null)

export const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false)

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    setAuthenticated(isAuthenticated)
  }, [])

  const login = useCallback((user) => {
    setAuthenticated(true)
    storage.set(AUTH_STORAGE_KEY, true)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  const logout = useCallback((user) => {
    setAuthenticated(false)
    storage.set(AUTH_STORAGE_KEY, false)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  )
}

// export { useAuth }
export default AuthProvider
