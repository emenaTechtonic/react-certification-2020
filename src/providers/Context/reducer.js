export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
