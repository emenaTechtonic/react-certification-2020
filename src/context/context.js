import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
