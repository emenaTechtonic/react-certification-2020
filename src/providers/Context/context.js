import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
