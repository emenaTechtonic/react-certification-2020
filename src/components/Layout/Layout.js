import React, { useContext } from 'react'
import Header from '../Header/Header'
import Menu from '../Menu/Menu'
import Container from '../Container/Container'
import { GlobalStyles } from '../../global'
import { Context } from '../../context/context'

import StyledLayout from './Layout.styled'

function Layout({ children }) {
  const { statusMenu, dark } = useContext(Context)

  return (
    <StyledLayout>
      <GlobalStyles statusMenu={statusMenu} theme={dark} />
      <Header />
      <Menu />
      <Container statusMenu={statusMenu}>{children}</Container>
    </StyledLayout>
  )
}

export default Layout
