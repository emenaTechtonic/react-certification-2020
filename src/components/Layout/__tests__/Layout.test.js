import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import { Context } from '../../../providers/Context/context'
import { AuthProvider } from '../../../providers/Auth/auth'
import Layout from '../Layout'

const state = {
  target: 'Wizeline',
  theme: false,
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
const login = null
const logout = null

describe('Testing Layout component', () => {
  test('renders Layout without crashing', async () => {
    const rendered = renderer.create(
      <AuthProvider authenticated={false} login={login} logout={logout}>
        <Context.Provider value={{ state }}>
          <MemoryRouter>
            <Layout />
          </MemoryRouter>
        </Context.Provider>
      </AuthProvider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
