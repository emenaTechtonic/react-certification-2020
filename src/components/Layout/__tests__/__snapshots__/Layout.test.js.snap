// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Layout component renders Layout without crashing 1`] = `
<div
  className="sc-pVTma cJPAIp layout"
>
  <div
    className="sc-jRQAMF jaGnlL header"
  >
    <div
      className="header__hamburger__search"
    >
      <div
        className="sc-gKckTs kqvsgi hamburger"
        onClick={[Function]}
      >
        <span />
        <span />
        <span />
      </div>
      <div
        className="sc-gsDJrp dLVeSJ"
      >
        <input
          className="sc-bdvvaa gzIPOt"
          onKeyDown={[Function]}
          placeholder="Wizeline"
          role="input"
          type="text"
        />
      </div>
    </div>
    <div
      className="header__theme__user"
    >
      <div
        className="sc-dkPtyc SvsMR toggle"
      >
        <input
          checked={false}
          className="checkbox"
          id="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          className="switch"
          htmlFor="checkbox"
        >
          <span
            className="slider"
          />
        </label>
        <div
          className="named"
        >
          Dark mode
        </div>
      </div>
      <div
        className="sc-hKwCoD emvDmI header__user"
      >
        <div
          className="popup login"
        >
          <ul>
            <li>
              Logout
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-iCfLBT FfNRE menu"
  >
    <a
      className="homeLink"
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <div />
  </div>
  <div
    className="sc-furvIG gXmMve container"
  />
</div>
`;
